monitoring/inotify_linux.go

```diff
@@ -0,0 +1,67 @@
+/*
+Copyright 2018 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitoring
+
+import (
+	"context"
+
+	"github.com/gravitational/satellite/agent/health"
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+	"github.com/gravitational/trace"
+
+	"golang.org/x/sys/unix"
+)
+
+// NewINotifyChecker creates a new health.Checker that tests if inotify watches
+// can be created. This is usually an indication that the system has reached
+// fs.inotify.max_user_instances has been exhausted
+func NewINotifyChecker() health.Checker {
+	return iNotifyChecker{}
+}
+
+// iNotifyChecker tries to create an inotify watch, and reports if it fails
+type iNotifyChecker struct {
+}
+
+// Name returns name of the checker
+func (c iNotifyChecker) Name() string {
+	return "inotify"
+}
+
+// Check tests whether inotify is working
+func (c iNotifyChecker) Check(ctx context.Context, reporter health.Reporter) {
+	fd, err := unix.InotifyInit()
```

Comment:
Can this return non-nil error and fd >= 0 (e.g. fd == 0)? If so, this case will be missed. I would modify the condition below to `if fd <0 || err != nil`, then all error cases should be covered.

Reply:
I don't believe this is possible, this is basically a direct mapping to the linux syscall, which used an FD of -1 to indicate error.